### PR TITLE
feat(activations): capture invocation type in state.json (DEV-78)

### PR DESCRIPTION
**Pull request by djsauble**
## Summary

- Adds `invocation_type: Option<InvocationType>` to `Attachment` so each attached PID's invocation kind (`Interactive`, `InPlace`, `ShellCommand`, `ExecCommand`) is recorded in `state.json` when present.
- Threads `invocation_type` through `ActivationState::start_or_attach` / `start` and preserves it across `replace_attachment` (no new `--invocation-type` arg on the `flox-activations attach` subcommand; the value is established once at activate time and carried forward).
- **Keeps schema at `Version<3>`.** The V4 bump is deferred to the DEV-82 PR that ships the consumer side, so users only see the "restart your activations" disruption once, paired with the actual feature.

Linear: [DEV-78](https://linear.app/floxdotdev/issue/DEV-78/capture-invocation-type-in-statejson)
Blocks: DEV-82

## Why the V4 bump is deferred

Bumping V3 → V4 in this PR alone would force every user to restart their activations on upgrade — for a field they can't yet use. The consumer (DEV-82, gated by `flox.features.auto_activate`) is not shipping in the same release. Per [Matthew's review feedback](https://github.com/flox/flox/pull/4248#discussion_r3251352952), deferring the bump means users see "here's a new feature, please restart your activations" rather than "please restart your activations now, the feature comes later."

The field is wrapped in `Option<InvocationType>` with `#[serde(default)]`. Both directions are safe:
- Older flox reading new `state.json`: serde drops the unknown key (no `#[serde(deny_unknown_fields)]` on `Attachment`).
- New flox reading older `state.json`: missing field deserializes as `None`.

New attachments created by this branch get `Some(invocation_type)`; `replace_attachment` preserves whatever value is there (`None` stays `None`). One new test, `deserialize_attachment_without_invocation_type_yields_none`, locks in the forward-compat contract for the duration of the deferral.

## Out of scope (handled in DEV-82)

- Consumer that reads `invocation_type` during `flox deactivate`.
- Replacing the `_FLOX_HOOK_DIRS` check in `handle_auto_deactivate`.
- Any change to `auto_activate` feature-flag gating itself.
- The `Version<3>` → `Version<4>` schema bump (see scope below).

## Scope of the eventual V4 bump (in the DEV-82 PR)

Mechanical, contained to two files:

1. `cli/flox-core/src/activations.rs`:
   - Drop `Option<>` and `#[serde(default)]` from `Attachment.invocation_type`.
   - Bump `ActivationState.version: Version<3>` → `Version<4>` and the `parse_versioned_activation_state` match arm.
   - Repoint `OLD_VERSION` (in `mod version_handling` tests) to `Version<3>`.
   - Unwrap `Some(...)` at every `Attachment { … }` literal.
   - Delete the `deserialize_attachment_without_invocation_type_yields_none` forward-compat test.
2. `cli/tests/activate.bats`: flip `incrementing_version_this_release` to `"true"`.

The migration tests for the V3 → V4 transition (running PIDs error / no-running-PIDs discard) get added there too, exercising the existing version-mismatch logic in `parse_versioned_activation_state`.

## Test plan

- [x] `nix develop -c just unit-tests` — 991 passed, 0 failed (includes the new forward-compat test)
- [x] `nix develop -c cargo clippy --workspace --all-targets` — clean
- [x] `nix develop -c cargo fmt --check` — clean
- [x] Forward-compat verified manually: new flox attaching to a `state.json` written by an older flox (no `invocation_type` field) deserializes existing attachments with `invocation_type: None` and adds its own with `Some(...)`.
- [ ] Manual: with stale `~/.cache/flox/run/activations` cleared, `flox activate` records `"invocation_type": "interactive"` for a subshell, `"inplace"` for `eval "$(flox activate -d <path>)"`, `"command"` for `flox activate -- echo hi`.
- [ ] `nix develop -c just integ-tests activate.bats`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: cli/flox-activations/src/cli/activate.rs [hunk 1]
```diff
@@ -137,7 +137,12 @@ impl ActivateArgs {
         let mut last_warning: Option<Instant> = None;
 
         loop {
-            match self.try_start_or_attach(context, subsystem_verbosity, vars_from_env)? {
+            match self.try_start_or_attach(
+                context,
+                invocation_type,
+                subsystem_verbosity,
+                vars_from_env,
+            )? {
```
Comment (djsauble) [inline cli/flox-activations/src/cli/activate.rs:145]: question: why is it necessary to pass through `invocation_type` when starting or attaching to an environment, when it wasn't before? Presumably Flox already knew what type of invocation was being performed, it just wasn't persisting it to `state.json`.

File: cli/flox-core/src/activations.rs [hunk 2]
```diff
@@ -677,6 +688,7 @@ impl ActivationState {
         let new_attachment = Attachment {
             start_id,
             expiration,
+            invocation_type: old_attachment.invocation_type,
```
Comment (djsauble) [inline cli/flox-core/src/activations.rs:695]: question: during the design phase, I was asked if I wanted an `--invocation-type` flag added to the CLI which could be used to specify any invocation type when attaching to an existing environment. Even though I said 'no' at the time and decided to default to the old attachment's invocation type, are there any situations where a user might want to choose a different invocation type, and why?

File: cli/flox-core/src/activations.rs [hunk 3]
```diff
@@ -878,6 +892,7 @@ mod tests {
         Attachment {
             start_id,
             expiration: None,
+            invocation_type: InvocationType::Interactive,
```
Comment (djsauble) [inline cli/flox-core/src/activations.rs]: question: why is `invocation_type` hardcoded to `InvocationType::Interactive` here? Are attachments only ever done while in interactive mode? Or is the test disregarding other possibilities?

Given that `invocation_type` wasn't passed into these tests before, what was the effective invocation type that was being tested before?

File: cli/flox-core/src/activations.rs [hunk 4]
```diff
@@ -1416,5 +1454,203 @@ mod tests {
...
+            let expected_msg = formatdoc! {"
+                This environment has already been activated with an incompatible version of 'flox'.
+
+                Exit all activations of the environment and try again.
+                PIDs of the running activations: {pid}",
+            };
+            assert_eq!(err.to_string(), expected_msg);
+
+            stop_process(proc);
+        }
```
Comment (djsauble) [inline cli/flox-core/src/activations.rs]: suggestion: I'm not sure this is worth testing. It's a very specific edge case that a given user will only ever encounter once (after upgrading from V3 to V4 schema) for each long-running environment they have.

File: cli/flox-core/src/activations.rs [hunk 5]
```diff
@@ -1416,5 +1454,203 @@ mod tests {
...
+                "attached_pids": {
+                    pid.to_string(): {},
+                },
+            })
+            .to_string();
+
+            let result = parse_versioned_activation_state(&json).unwrap();
+            assert_eq!(result, None, "should discard pre-V4 state");
+        }
+    }
```
Comment (djsauble) [inline cli/flox-core/src/activations.rs]: thought: even though this test is a similar edge case to the one above, it may be worth keeping this just to ensure that the V3 schema is being correctly discarded once there are no longer any long-running activations with older versions of Flox.

File: cli/flox-core/src/activations.rs [hunk 6]
```diff
@@ -1416,5 +1454,203 @@ mod tests {
...
+        #[test]
+        fn attachment_roundtrips_exec_command() {
+            let attachment = attachment_with(InvocationType::ExecCommand(vec![
+                "ls".to_string(),
+                "-l".to_string(),
+            ]));
+            let json = serde_json::to_string(&attachment).unwrap();
+            let parsed: Attachment = serde_json::from_str(&json).unwrap();
+            assert_eq!(parsed, attachment);
+        }
```
Comment (djsauble) [inline cli/flox-core/src/activations.rs]: suggestion: these tests don't seem to be providing much value. They're just checking whether or not an `Attachment` can be converted to a string and back without changing. Unless I'm missing something I'd say we can just remove these, the following tests seem more useful.

File: cli/flox-core/src/activations.rs [hunk 7]
```diff
@@ -1416,5 +1454,203 @@ mod tests {
...
+
+            let pid = 123;
+            let result = activations.start_or_attach(
+                pid,
+                &store_path,
+                InvocationType::ShellCommand("echo hi".to_string()),
+            );
+            let start_id = match result {
+                StartOrAttachResult::Start { start_id } => start_id,
+                _ => panic!("Expected Start, got {:?}", result),
```
Comment (djsauble) [inline cli/flox-core/src/activations.rs]: question: is `panic!` an idiomatic way to fail the test?

File: cli/flox-core/src/activations.rs [hunk 8]
```diff
@@ -1416,5 +1454,203 @@ mod tests {
...
+                .replace_attachment(start_id.clone(), pid, pid, Some(expiration))
+                .unwrap();
+
+            let expected = Attachment {
+                start_id,
+                expiration: Some(expiration),
+                invocation_type: InvocationType::ExecCommand(vec!["ls".to_string()]),
+            };
+            assert_eq!(activations.attached_pids, BTreeMap::from([(pid, expected)]));
+        }
```
Comment (djsauble) [inline cli/flox-core/src/activations.rs]: question: In what situations would an attachment get replaced with an attachment that has an expiration? And just to confirm, we're not actually testing the 60 second timeout itself, right? I don't want to inadvertently add a 60 second delay to the test suite.

File: cli/flox-core/src/activations.rs [hunk 9]
```diff
@@ -1416,5 +1454,203 @@ mod tests {
...
+            let expected = Attachment {
+                start_id,
+                expiration: None,
+                invocation_type: InvocationType::InPlace,
+            };
+            assert_eq!(
+                activations.attached_pids,
+                BTreeMap::from([(new_pid, expected)])
+            );
+        }
```
Comment (djsauble) [inline cli/flox-core/src/activations.rs:1514]: question: Similarly, in what situations would an attachment have its PID replaced with a different PID? Just trying to understand if this is a real thing that happens that needs a test.

File: cli/flox-activations/src/cli/activate.rs (hunk 1 above)
Comment (djsauble) [reply cli/flox-activations/src/cli/activate.rs:145]: Right — Flox always knew the type at the call site; the difference is that `Attachment` now persists it, so something has to set it when an attachment is created. It flows through as a parameter (rather than being read from a global) because **invocation type is per-attachment, not per-activation**: one ready activation can have multiple attached PIDs from different routes — e.g. an interactive shell, then a sibling `flox activate -- echo hi` (`ShellCommand`) attaching to the same `start_id`. Each row in `attached_pids` records how *that* PID arrived; we can't infer it from the start.

File: cli/flox-core/src/activations.rs (hunk 2 above)
Comment (djsauble) [reply cli/flox-core/src/activations.rs:695]: Looking at the two callers (`flox-activations attach --timeout-ms` and `--remove-pid`), both are internal bookkeeping from the in-place activation flow:

1. `flox activate` records itself with `invocation_type = InPlace`.
2. `flox activate` calls `attach --timeout-ms 5000` against its own PID so the activation isn't GC'd before the parent shell evaluates the printed script.
3. The parent shell evaluates the script, which calls `attach --remove-pid <flox-pid>` to hand ownership to the long-lived shell.

The logical invocation type doesn't change between steps — same in-place activation; the PID changes hands and gains an expiration, but `eval $(flox activate)` is fixed at start time. Letting the caller specify a new type would just create a way for the bash glue to lie about how the activation was started. If a future caller did want differing types, the right knob would still be at `start_or_attach` time, not `replace_attachment`.

File: cli/flox-core/src/activations.rs (hunk 3 above)
Comment (djsauble) [reply cli/flox-core/src/activations.rs]: `make_attachment` is used by tests in `mod read_and_write_state`, `mod attached_pids_getters`, `mod start_or_attach`, and `mod version_handling` — they exercise the state machine (start/attach/ready/cleanup) and shouldn't care which `InvocationType` they carry. We need *some* value; `Interactive` is the most representative default. Tests that *do* care about the field live in `mod invocation_type` and pass explicit values.

> What was the effective invocation type tested before?

None. The field didn't exist before this PR, so the older tests are still invocation-type-agnostic; this helper just supplies a placeholder so they keep compiling.

File: cli/flox-core/src/activations.rs (hunk 4 above)
Comment (djsauble) [reply cli/flox-core/src/activations.rs]: You're right — the existing OLD_VERSION-based tests already cover this code path. Looking closer: `OLD_VERSION = Version<2>` was always synthetic (V1/V2 lived at a different state.json path and aren't reachable through `parse_versioned_activation_state`). Now that V3 is the realistic previous version, I repointed `OLD_VERSION` to `Version<3>` and dropped both new V3-specific tests as redundant. The existing OLD_VERSION tests now describe the real V3 → V4 transition with broader coverage (running attachments, no running attachments, executive-only, malformed, structurally incompatible).

Done in 748d340fc.

File: cli/flox-core/src/activations.rs (hunk 5 above)
Comment (djsauble) [reply cli/flox-core/src/activations.rs]: Tied to the previous comment — rather than keeping this in isolation, I repointed `OLD_VERSION` to `Version<3>` so `parse_versioned_activation_state_different_version_pids_not_running` now describes the realistic V3 discard path. Net: zero duplication, broader coverage retained.

Done in 748d340fc.

File: cli/flox-core/src/activations.rs (hunk 6 above)
Comment (djsauble) [reply cli/flox-core/src/activations.rs]: Agreed. `#[derive(Serialize, Deserialize)]` round-tripping is `serde`'s job. The integration is already covered: `read_and_write_roundtrip` exercises full disk serialize/deserialize, and `start_or_attach_records_invocation_type_on_{start,attach}` assert equality on the whole `Attachment` after going through the state machine (a busted `Serialize` would fail them indirectly). Removed all four `attachment_roundtrips_*` tests and the now-unused `attachment_with` helper.

Done in 748d340fc.

File: cli/flox-core/src/activations.rs (hunk 7 above)
Comment (djsauble) [reply cli/flox-core/src/activations.rs]: Yes. The file already uses `_ => panic!("Expected …, got {:?}", result)` extensively (lines 967, 1047, 1052, 1162, …). The other idiom in the same PR is `let-else { panic!(…) }` (e.g. `attach.rs:125–127`). Both are idiomatic Rust for tests; `panic!` with a descriptive message gives a clear failure. The unstable `assert_matches!` macro is an alternative but isn't in std. Keeping the existing style to match the surrounding file.

File: cli/flox-core/src/activations.rs (hunk 8 above)
Comment (djsauble) [reply cli/flox-core/src/activations.rs]: Two parts:

**When this happens:** In-place activation. See `cli/flox-activations/src/attach.rs:526` (`activate_in_place`) — the short-lived `flox activate` process records itself, then calls `AttachArgs { timeout_ms: Some(5000), … }` against its own PID. That keeps the activation alive long enough for the parent shell to evaluate the printed script before cleanup. This test covers the same `replace_attachment(self_pid, self_pid, Some(expiration))` shape.

**About the 60s:** No real wait. `expiration = now + Duration::from_secs(60)` is just a sentinel to check that the field round-trips through `replace_attachment` unchanged. Nothing sleeps; the test runs in well under a second.

File: cli/flox-core/src/activations.rs (hunk 9 above)
Comment (djsauble) [reply cli/flox-core/src/activations.rs:1514]: Real and frequent — same in-place activation flow (`cli/flox-activations/src/attach.rs:571`):

```
{flox_activations} attach … --remove-pid "{pid}"
```

is emitted into the script `flox activate` prints for the parent shell to evaluate. `--remove-pid` triggers `replace_attachment(start_id, old_pid=flox_activate_pid, new_pid=shell_pid, None)`, swapping ownership from the short-lived flox process to the long-lived shell. The existing `process_exited_noop_for_pid_not_in_state` test (`cli/flox-activations/src/cli/executive/mod.rs:629`) exercises the same mechanism from a different angle.

File: cli/flox-activations/src/cli/executive/mod.rs [hunk 10]
```diff
@@ -514,7 +515,7 @@ mod test {
         // Create an ActivationState with one PID attached
         let mut state =
             ActivationState::new(&ActivateMode::default(), Some(&dot_flox_path), &flox_env);
-        let result = state.start_or_attach(pid, &store_path);
```
Comment (mkenigs) [inline cli/flox-activations/src/cli/executive/mod.rs:517]: (unrelated line) suggestion blocking:
> Storage is unconditional; the auto_activate feature
flag gates only the consumer side

I'm not sure if it will be tricky, but I think we want to gate the schema bump with an actual feature. I think it's better for users if we say "here's this new feature, go restart your activations" rather than "go restart your activations" and then release the feature later.

If that ends up being too difficult we could decide to bump it now, but at least worth investigating if we can feature flag it

Comment (djsauble) [reply cli/flox-activations/src/cli/executive/mod.rs:517]: That's a great point. I think we can pull this off.

Comment (djsauble) [reply cli/flox-activations/src/cli/executive/mod.rs:517]: In https://github.com/flox/flox/pull/4248/commits/1a3715a525c9becf839dee2ce43b86c04d321b22 I kept the schema version the same and made `invocation_type` an optional field. And it seems to work with both forwards and backwards compatibility.

1. If I do a `flox activate` on an older version of Flox, and then attach to the same environment using the new version, I see that `invocation_type` is only set in `state.json` for the new version (and it's `null` for the older version):

```
"attached_pids": {
  "23196": {
    "expiration": null,
    "invocation_type": null,
    "start_id": {
      "store_path": "/nix/store/lnrbfb2ixviylf719c91qb6i6i0l3akj-environment-develop",
      "timestamp": 1778887713624
    }
  },
  "23316": {
    "expiration": null,
    "invocation_type": "interactive",
    "start_id": {
      "store_path": "/nix/store/lnrbfb2ixviylf719c91qb6i6i0l3akj-environment-develop",
      "timestamp": 1778887713624
    }
  }
},
```
  
2. If I swap the order and run `flox activate` on the newer version of Flox first, and then attach using the older version, I see that `invocation_type` isn't set in `state.json` at all.
  
```
"attached_pids": {
  "23499": {
    "expiration": null,
    "start_id": {
      "store_path": "/nix/store/lnrbfb2ixviylf719c91qb6i6i0l3akj-environment-develop",
      "timestamp": 1778887816842
    }
  },
  "23608": {
    "expiration": null,
    "start_id": {
      "store_path": "/nix/store/lnrbfb2ixviylf719c91qb6i6i0l3akj-environment-develop",
      "timestamp": 1778887816842
    }
  }
},
```
  
Once we're ready to ship DEV-82, we can bump the schema version to V4 and make `invocation_type` a required field.

Comment (mkenigs) [reply cli/flox-activations/src/cli/executive/mod.rs:517]: > If I swap the order and run flox activate on the newer version of Flox first, and then attach using the older version, I see that invocation_type isn't set in state.json at all

Looks like it's set by the `flox activate` of the newer Flox and then the file is re-written without it by the older version. Seems okay 👍

File: cli/flox-core/src/activations.rs [hunk 11]
```diff
@@ -1417,4 +1457,124 @@ mod tests {
...
+            let expected = Attachment {
+                start_id,
+                expiration: None,
+                invocation_type: Some(InvocationType::InPlace),
+            };
+            assert_eq!(activations.attached_pids, BTreeMap::from([(pid, expected)]));
+        }
+
+        #[test]
+        fn replace_attachment_preserves_invocation_type_on_timeout() {
```
Comment (mkenigs) [inline cli/flox-core/src/activations.rs]: suggestion blocking: I don't think timeout is actually reached here, so I'm not sure what `on_timeout` means. If we have coverage for timeouts above that asserts on a `make_attachment(...)` we may have coverage for this case already (see other review comment)

File: cli/flox-core/src/activations.rs [hunk 12]
```diff
@@ -1417,4 +1457,124 @@ mod tests {
             stop_process(exec_proc);
         }
     }
+
+    mod invocation_type {
+        use super::*;
+
+        #[test]
+        fn start_or_attach_records_invocation_type_on_start() {
```
Comment (mkenigs) [inline cli/flox-core/src/activations.rs]: suggestion nonblocking: I think this is already covered by `test_start_or_attach_starts_when_ready_false` because that asserts
```
            assert_eq!(
                activations.attached_pids,
                BTreeMap::from([(pid, make_attachment(start_id))])
            );
```
and `make_attachment` includes an invocation type

Sometimes it's nice to have more targeted unit tests rather than mega-tests that assert everything, but since it's already covered and that test isn't too mega I'd lean towards dropping this test

Similar comment for `start_or_attach_records_invocation_type_on_attach` and `start_or_attach_records_invocation_type_on_attach` (the latter is covered by `test_start_or_attach_replaces_existing_pid`)

**Review by mkenigs — CHANGES_REQUESTED** [review]

File: cli/flox-activations/src/cli/executive/mod.rs (hunk 10 above)
Comment (djsauble) [reply cli/flox-activations/src/cli/executive/mod.rs:517]: Yep, that's correct.

File: cli/flox-core/src/activations.rs (hunk 12 above)
Comment (djsauble) [reply cli/flox-core/src/activations.rs]: Removes `*_on_start` and `*_on_attach`. https://github.com/flox/flox/pull/4248/changes/a7409ed25d4c223b99b1fd80009643c96acb5818

File: cli/flox-core/src/activations.rs (hunk 11 above)
Comment (djsauble) [reply cli/flox-core/src/activations.rs]: The `replace_attachment_preserves_invocation_type_on_timeout` test was badly named. It's actually testing the [`replace_attachment` code path](https://github.com/flox/flox/pull/4248/changes#diff-58e7d2da3cb9b6714f5b9afd187f2a87ae73a45ec4f9a4008fa4266b3a21d34cR676-R701) (not `make_attachment`) with an expiration date. I renamed the test here to make this clearer. https://github.com/flox/flox/pull/4248/changes/d58ac49b46b6ed55664cc94ce1a9e56609ebaef7

(The related `*_on_pid_swap` test is testing the same code path but exercising a PID change instead of expiration date).

I can still remove them if you want, but that's the context.

Comment (mkenigs) [reply cli/flox-core/src/activations.rs]: I'd have to dig through the `start_or_attach` tests to figure out whether these two tests are repeating existing coverage, but not a huge deal, so nonblocking/I'll leave it up to you

**Review by mkenigs — APPROVED** [review]
